src/vs/vscode.proposed.d.ts

```diff
@@ -1554,24 +1564,76 @@ declare module 'vscode' {
 	}
 
 	export interface NotebookProvider {
-		resolveNotebook(editor: NotebookEditor): Promise<void>;
-		executeCell(document: NotebookDocument, cell: NotebookCell | undefined): Promise<void>;
+		resolveNotebook(editor: NotebookEditor): Promise<void>; // rchiodo: Not sure what this is for? Notebook is loaded? Like prime the server? Or does this somehow generate the document?
 		save(document: NotebookDocument): Promise<boolean>;
 	}
 
+	/**
+	 * Result of an executeCell.
+	 */
+	export interface NotebookCellExecutionResult {
+		/**
+		 * Event fired when new cell output comes through
+		 */
+		readonly onCellOutput: Event<CellOutput[]>;
+		/**
+		 * Event fired when output should be cleared. Boolean flag indicates if should wait for next onOutput for clear.
+		 */
+		readonly onClear: Event< boolean>;
+		/**
+		 * Event fired when document wide output comes through
+		 */
+		readonly onDocumentOutput: Event<CellOutput[]>;
+		/**
+		 * Event fired when execution is complete
+		 */
+		readonly onComplete: Event<void>;
+	}
+
+	/**
+	 * Allows execution of notebook cells.
+	 */
+	export interface NotebookExecution {
+		readonly executionCount: number;
+		executeCell(document: NotebookDocument, cell: NotebookCell): NotebookCellExecutionResult;
+	}
+
+	/**
+	 * Allows observing the execution of cells from a NotebookExecution
+	 */
+	export interface NotebookExecutionObserver {
+		onExecuteStart(document: NotebookDocument, cell: NotebookCell): void;
+		onCellOutput(document: NotebookDocument, cell: NotebookCell, outputs: CellOutput[]): void;
+		onCellClear(document: NotebookDocument, cell: NotebookCell, wait: boolean): void;
+		onDocumentOutput(document: NotebookDocument, cell: NotebookCell, outputs: CellOutput[]): void;
+		onExecuteComplete(document: NotebookDocument, cell: NotebookCell, executionCount: number): void;
+	}
+
 	export interface NotebookOutputSelector {
 		type: string;
 		subTypes?: string[];
 	}
 
+	/**
+	 * Allows transformation of output from raw execution. Called during execution.
+	 */
+	export interface NotebookOutputTransformer {
+		/**
+		 *
+		 * @returns the next item in the chain
+		 *
+		 */
+		transformOutput(document: NotebookDocument, cell: NotebookCell, output: CellOutput): CellOutput;
```

Comment:
Isn't the the responsibility of a rendered?
I.e. extension authors can do their transformation and return the result in `NotebookOutputRender`.
Making it unnecessary for VSC to ask for extensions to transform output, when VSC is only ever concerned with the `string output` for rendering.

Reply:
No this transforms one output type to another. It doesn't render anything.

Reply:
Say given a image, it turns it into an ipywidget request (although not sure how that's going to work).

Reply:
Yeah how would ipywigets work at all with this model. They need the comm with the kernel. I guess we'd have to put them inside the same extension that implements execution

Reply:
The use case for the transform is actually for our error line modification. It transforms an error into something with links in it.

Reply:
But you're right we could just have a text/error handler.

Reply:
> Say given a image, it turns it into an ipywidget request (although not sure how that's going to work).

Again, the only purpose of this is to render something, Hence this is satisfied with what we have today.

Reply:
> They need the comm with the kernel. I guess we'd have to put them inside the same extension that implements execution

We use `preloads` and then render the result. The render result contains `javascript` that has the necessary context (loaded from preload), that can then communicate back, to extension with `onMessage` or some other mechanism. Cuz VSC cannot use `onCellExecuted` at all, cuz IPYWidgets needs to be able to send/receive a stream and current VSC API wpont' work and i think its overkill to use that interface to transmit ipywidget data back and forth.

Reply:
> It transforms an error into something with links in it.

So why don't we have our own result that renders this.
I wouldn't want other extensions coming and meddling with the `<html>` result that I generated for display in VSC.

That could result in unexpected/unwanted results. I.e. one extension modifying the Html provided by another extension.

Reply:
Yeah maybe, I want to leave it there to see what Kai thinks. The transform pipeline was his idea.